### PR TITLE
UIP-49 First steps on narrativemanager updating.

### PR DESCRIPTION
**Pull request by briehl**
First steps here for the NarrativeManager module fixup. These include:
1. pytest-ifying the unit tests
2. add the "workspace mock" (which really makes a mini internal workspace, sort of, which I think is entirely overkill and leads to being some weird hybrid between unit testing and integration testing) library to the conftest fixtures.
3. Do linting and typing on the narrativemanager module. I did my best to not change any actual code here, as I want a more complete regression test suite first before I go poking around and undoing things that should've been changed 5+ years ago.

### CODE REVIEW

File: lib/NarrativeService/NarrativeServiceImpl.py [hunk 1]
```diff
@@ -41,8 +41,6 @@ def _nm(self, ctx):
         """
         return NarrativeManager(self.config,
                                 ctx["user_id"],
-                                self._get_set_api_client(ctx["token"]),
-                                self._get_data_palette_client(ctx["token"]),
```
Comment (briehl) [inline lib/NarrativeService/NarrativeServiceImpl.py:45]: These aren't actually used anywhere, so removed from the narrative manager module.

File: test/conftest.py [hunk 2]
```diff
@@ -56,7 +58,7 @@ def auth_client(config: dict[str, str | int]) -> Generator[KBaseAuth, None, None
     yield KBaseAuth(config["auth-service-url"])
 
 @pytest.fixture(scope="module")
-def context(auth_token: str, auth_client: KBaseAuth) -> Generator[dict[str, any], None, None]:
+def context(auth_token: str, auth_client: KBaseAuth) -> Generator[dict[str, Any], None, None]:
```
Comment (briehl) [inline test/conftest.py:61]: `any` is a builtin function. `typing.Any` is the type. Oops. (more in the next PR).

File: test/unit/test_narrativemanager.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
+"""
```
Comment (briehl) [inline test/unit/test_narrativemanager.py:1]: Gave this the pytest treatment, but mostly left it unchanged.

File: lib/NarrativeService/narrativemanager.py
Comment (briehl) [inline lib/NarrativeService/narrativemanager.py:1]: I tried really hard not to change any actual code here. Most of the changes are either linting, typing, or renaming variables from camelCase to snake_case. I'll point out any real changes.

File: lib/NarrativeService/narrativemanager.py [hunk 4]
```diff
@@ -21,24 +25,32 @@ class NarrativeManager:
...
+    def __init__(
+        self,
+        config: dict[str, Any],
+        user_id: str,
+        workspace_client: Workspace,
+        search_service_client: SearchServiceClient
+    ) -> None:
         self.narrativeMethodStoreURL = config["narrative-method-store"]
-        self.set_api_client = set_api_client                     # DynamicServiceCache type
-        self.data_palette_client = data_palette_client          # DynamicServiceCache type
```
Comment (briehl) [inline lib/NarrativeService/narrativemanager.py:27]: These are set, then never used by the module. So they've been removed.

File: lib/NarrativeService/narrativemanager.py [hunk 5]
```diff
@@ -164,261 +178,321 @@ def revert_narrative_object(self, obj):
...
+    def _check_new_version_indexed(
+        self,
+        obj: dict[str, int | str],
+        new_version: int
+    ) -> dict[str, Any]:
         tries = 0
+        max_tries: int = 60
         data = None
-        while tries < 60 and data is None:
+        while tries < max_tries and data is None:
```
Comment (briehl) [inline lib/NarrativeService/narrativemanager.py:189]: Still a "magic number" but passes linting.
A future PR will clean this up further, and likely export it to a different module.

File: lib/NarrativeService/narrativemanager.py (hunk 4 above)
Comment (ialarmedalien) [reply lib/NarrativeService/narrativemanager.py:27]: 😖

File: lib/NarrativeService/narrativemanager.py [hunk 6]
```diff
@@ -21,24 +25,32 @@ class NarrativeManager:
     KB_CODE_CELL = "kb_code"
     KB_STATE = "widget_state"
 
-    def __init__(self, config, user_id, set_api_client, data_palette_client, workspace_client, search_service_client):
+    def __init__(
+        self,
```
Comment (ialarmedalien) [inline lib/NarrativeService/narrativemanager.py:29]: The `self`s can all have the class `"NarrativeManager"` (with quotes)

File: lib/NarrativeService/narrativemanager.py [hunk 7]
```diff
@@ -164,261 +178,321 @@ def revert_narrative_object(self, obj):
...
-    def copy_narrative(self, newName, workspaceRef, workspaceId):
+    def copy_narrative(
+        self,
+        new_name: str,
+        workspace_ref: str,
+        workspace_id: int
+    ) -> dict[str, str]:
         time_ms = int(round(time.time() * 1000))
-        newWsName = self.user_id + ":narrative_" + str(time_ms)
+        new_ws_name = self.user_id + ":narrative_" + str(time_ms)
```
Comment (ialarmedalien) [inline lib/NarrativeService/narrativemanager.py:214]: I love that it suddenly starts using camel case halfway through the file!

File: test/unit/test_narrativemanager.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    # ensure that there are no kbase narrative instances in data objects
+    for data_obj in doc["data_objects"]:
+        assert NARRATIVE_TYPE not in data_obj["obj_type"]
+
+
+def test_get_narrative_doc_bad_upa(config, mock_workspace_client, mock_user):
+    nm = NarrativeManager(config, mock_user, mock_workspace_client, mock.MagicMock())
+    # test that poorly formatted upa is handled correctly
+    with pytest.raises(ValueError, match="required format is <workspace_id>/<object_id>/<version>"):
+        nm.get_narrative_doc("blah/blah/blah/blah")
```
Comment (ialarmedalien) [inline test/unit/test_narrativemanager.py:64]: what happens if the UPA is in the format `ws_name/obj_name`?

For these tests, is it worth creating a test workspace with the various needed objects and then recording the responses with vcrpy?

File: test/unit/test_narrativemanager.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        mock_user,
+        make_object_history=True
+    )
+
+    (ws_id, obj, _) = narrative_ref.split("/")
+
+    nm = NarrativeManager(config, mock_user, mock_workspace_client, mock.MagicMock())
+
+    # simulate reverting fake narrative to version #2
+    # (make_object_history=True automatically makes 5 versions)
```
Comment (ialarmedalien) [inline test/unit/test_narrativemanager.py:88]: could you put this comment up next to where you create the narrative?

File: test/unit/test_narrativemanager.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+def test_revert_narrative_object(config, mock_workspace_client, mock_user):
+    # set up narrative
+    narrative_ref = mock_workspace_client.make_fake_narrative(
+        "SomeNiceName",
+        mock_user,
+        make_object_history=True
+    )
+
+    (ws_id, obj, _) = narrative_ref.split("/")
```
Comment (ialarmedalien) [inline test/unit/test_narrativemanager.py:83]: worth checking the version of the narrative ref here to ensure there are five versions?

File: lib/NarrativeService/narrativemanager.py (hunk 7 above)
Comment (briehl) [reply lib/NarrativeService/narrativemanager.py:214]: The person who originally wrote this (and other chunks of NarrativeService) was primarily a Java programmer. Also, this particular file was basically directly translated from a JavaScript module. I guess naming conventions came along for the ride.

File: test/unit/test_narrativemanager.py (hunk 8 above)
Comment (briehl) [reply test/unit/test_narrativemanager.py:64]: > is it worth creating a test workspace with the various needed objects?

I think that's worth doing for integration tests, once those get reworked. I think the issue here is that we're testing modifications to existing data. For the unit tests, I think I want to just test the NarrativeManager code in question - does it fire off the right backend call ok and format results properly. This currently does kind of a weird hodge-podge that, IMO, isn't necessary.

Integration tests should go all in, make a new Workspace, put data in it, modify it, see if the Workspace service responds right, changes the data as expected, and so on.

Comment (briehl) [reply test/unit/test_narrativemanager.py:64]: Mostly, I'd like to just pytest + ruff in this PR, keep all the warts the same for consistency, then go back in the next PR and rework the process a bit.

**Comment by briehl** [comment]
Responded to comments, and made a few requested changes.

**Review by ialarmedalien — APPROVED** [review]

**Pull request merged**